File: TicTacToe.kt
package tictactoe

fun main() {
    var runs = 0
    var endGame = false
    val a = mutableListOf(" ", " ", " ", " ", " ", " ", " ", " ", " ")

    fun printGraph(a: MutableList<String>){
        println(
            """
            ---------
            | ${a[0]} ${a[1]} ${a[2]} |
            | ${a[3]} ${a[4]} ${a[5]} |
            | ${a[6]} ${a[7]} ${a[8]} |
            ---------
            """.trimIndent()
        )
    }
    printGraph(a)

    game@while (true) {
        val b = readln().uppercase().filter { !it.isWhitespace() }
        val x = b[0].toString().toInt() - 1
        val y = b[1].toString().toInt() - 1
        val p = 3 * x + y

        when {
            !b.all { it in '0'..'9' } -> println("You should enter numbers!")
            x > 2 || y > 2 -> println("Coordinates should be from 1 to 3!")
            a[p] == "X" || a[p] == "O" -> println("This cell is occupied! Choose another one!")
            else -> {  // run the game
                a[p] = if (runs++ % 2 == 0) "X" else "O"
                printGraph(a)
                fun check(d: String = "") {
                    when (d) {
                        "X" -> println("X wins")
                        "O" -> println("O wins")
                        else -> println("Draw")
                    }
                    endGame = true
                }

                when {
                    (a[0] == a[1] && a[1] == a[2] && a[0] != " ") -> check(a[0]) // top   row
                    (a[3] == a[4] && a[4] == a[5] && a[3] != " ") -> check(a[3]) // mid   row
                    (a[6] == a[7] && a[7] == a[8] && a[6] != " ") -> check(a[6]) // bot   row
                    (a[0] == a[3] && a[3] == a[6] && a[0] != " ") -> check(a[0]) // left  col
                    (a[1] == a[4] && a[4] == a[7] && a[1] != " ") -> check(a[1]) // mid   col
                    (a[2] == a[5] && a[5] == a[8] && a[2] != " ") -> check(a[2]) // right col
                    (a[0] == a[4] && a[4] == a[8] && a[4] != " ") -> check(a[4]) // diagonals \
                    (a[2] == a[4] && a[4] == a[6] && a[4] != " ") -> check(a[4]) // diagonals /
                    runs == 9 -> check()  //draw
                }
                if (endGame) break@game
            }
        }
    }
}

File: Python/python_project/TicTacToe.py
#!/usr/bin/env python3

def print_grid(g):
    print("---------")
    for i in g:
        print(f'| {" ".join(i)} |')
    print("---------")
    return


def move(g):
    steps = 0

    while True:
        step = input().split()
        if step[0].isnumeric() and step[1].isnumeric():
            a = int(step[0]) - 1
            b = int(step[1]) - 1
            if not 0 <= a <= 2 or not 0 <= b <= 2:
                print("Coordinates should be from 1 to 3!")
            elif g[a][b] != "X" and g[a][b] != "O":
                if steps % 2 == 0:
                    g[a][b] = "X"
                else:
                    g[a][b] = "O"
                steps += 1
                ls = ["".join(g[0]), "".join(g[1]), "".join(g[2]),
                      g[0][0] + g[1][0] + g[2][0], g[0][1] + g[1][1] + g[2][1],
                      g[0][0] + g[1][1] + g[2][2], g[0][2] + g[1][1] + g[2][0]]
                print_grid(g)
                if "XXX" in ls:
                    print("X wins")
                    return
                elif "OOO" in ls:
                    print("O wins")
                    return
                elif steps == 9:
                    print("Draw")
                    return
            else:
                print("This cell is occupied! Choose another one!")
        else:
            print("You should enter numbers!")


def main():
    grid = [
        [" ", " ", " "],
        [" ", " ", " "],
        [" ", " ", " "]
    ]
    print_grid(grid)
    move(grid)


main()
